### PR TITLE
EHN: auto retry download on network errors

### LINKED ISSUE #302: ENH: auto retry download on network errors
### Is your feature request related to a problem? Please describe
#301 describes a common problem when downloading large models.

### Describe the solution you'd like
Auto retry on network errors. Enable`resume_download` so that the previous progress won't be lost.

### PR DESCRIPTION
**Pull request by jiayini1119**
Resolves #302.

### CODE REVIEW

**Comment by UranusSeven** [comment]
We should retry on [LocalEntryNotFoundError](https://huggingface.co/docs/huggingface_hub/package_reference/utilities#huggingface_hub.utils.LocalEntryNotFoundError).

Catch the exception above and download again with same parameters should work.

File: xinference/model/llm/llm_family.py [hunk 1]
```diff
@@ -226,25 +226,49 @@ def cache_from_huggingface(
...
+                    local_dir=cache_dir,
+                    local_dir_use_symlinks=True,
+                )
+                break
+            except huggingface_hub.utils.LocalEntryNotFoundError:
+                pass
+            except Exception as e:
+                logger.error(f"fail to download the model: {e}")
+                break
+        else:
```
[the remark this answers is not in the document]
Comment (jiayini1119) [reply xinference/model/llm/llm_family.py:270]: Here I'm using the [for-else](https://book.pythontips.com/en/latest/for_-_else.html) structure. The else block is executed when the for loop runs to completion without encountering a break statement.

File: xinference/model/llm/llm_family.py [hunk 2]
```diff
@@ -228,22 +230,53 @@ def cache_from_huggingface(
...
+                    revision=llm_spec.model_revision,
+                    filename=file_name,
+                    local_dir=cache_dir,
+                    local_dir_use_symlinks=True,
+                )
+                break
+            except huggingface_hub.utils.LocalEntryNotFoundError:
+                remaining_attempts = MAX_RETRIES - current_attempt
+                logger.warning(
+                    f"Attempt {current_attempt} failed. Max remaining attempts: {remaining_attempts}"
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:272]:
```suggestion
                    f"Attempt {current_attempt} failed. Remaining attempts: {remaining_attempts}"
```

File: xinference/model/llm/llm_family.py [hunk 3]
```diff
@@ -228,22 +230,53 @@ def cache_from_huggingface(
...
+                    llm_spec.model_id,
+                    revision=llm_spec.model_revision,
+                    local_dir=cache_dir,
+                    local_dir_use_symlinks=True,
+                )
+                break
+            except huggingface_hub.utils.LocalEntryNotFoundError:
+                remaining_attempts = MAX_RETRIES - current_attempt
+                logger.warning(
+                    f"Attempt {current_attempt} failed. Max remaining attempts: {remaining_attempts}"
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:246]:
```suggestion
                    f"Attempt {current_attempt} failed. Remaining attempts: {remaining_attempts}"
```

File: xinference/model/llm/llm_family.py [hunk 4]
```diff
@@ -228,22 +230,53 @@ def cache_from_huggingface(
...
+                )
+                pass
+            except Exception as e:
+                raise RuntimeError(f"Failed to download the model: {e}")
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:276]:
```suggestion
```

File: xinference/model/llm/llm_family.py [hunk 5]
```diff
@@ -228,22 +230,53 @@ def cache_from_huggingface(
...
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:250]:
```suggestion
```

File: xinference/model/llm/llm_family.py [hunk 6]
```diff
@@ -26,6 +26,8 @@
 
 logger = logging.getLogger(__name__)
 
+MAX_RETRIES = 3
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:29]: This constant actually means max attempts.

File: xinference/model/llm/llm_family.py [hunk 7]
```diff
@@ -228,22 +230,49 @@ def cache_from_huggingface(
...
+                    local_dir=cache_dir,
+                    local_dir_use_symlinks=True,
+                )
+                break
+            except huggingface_hub.utils.LocalEntryNotFoundError:
+                remaining_attempts = MAX_RETRIES - current_attempt
+                logger.warning(
+                    f"Attempt {current_attempt} failed. Remaining attempts: {remaining_attempts}"
+                )
+                pass
```
Comment (UranusSeven) [inline xinference/model/llm/llm_family.py:248]: Is `pass` necessary?

**Review by UranusSeven — APPROVED** [review]

**Pull request merged**